terms/fast_fourier_transforms.md

```diff
@@ -1,3 +1,153 @@
+def inverse_fft(vals, modulus, domain):
+    vals = fft(vals, modulus, domain)
+    return [x * modular_inverse(len(vals), modulus) % modulus for x in [vals[0]] + vals[1:][::-1]]
+```
+
+We can use FFT to multiply numbers in time $O(NlogN)$ which $N$ is the length of numbers, faster than $O(N^2)$ schoolbook
+multiplication.
+
+## FFTs and Binary Fields
+
+In an binary field, each element is expressed as a polynomial where all of the entries are 0 or 1, eg. $x^3+x+1$. Adding polynomials
+is done modulo 2, and subtraction is the same as addition (as $−1=1 \mod2$). We select a irreducible polynomial as a modulus
+(e.g., $x^4 + x + 1$) and multiplication is done modulo that modulus.
+
+If you want to run FFT in a binary field, there is a problem:
+In binary fields, there are *no multiplicative groups of order $2^n$* because all the multiplicative groups are order $2^n - 1$.
+
+For example, in the binary field with modulus $x^4+x+1$, if you start calculating successive powers of $x+1$, you cycle back to $1$
+after $15$ steps - not $16$. The reason is that the elements list includes zero, but you're never going to reach zero by multiplying
+any nonzero value by itself in a field, so the power of $x+1$ cycle through every element but zero.
+
+So how can we halve the size of a domain? Here is the solution.
+
+Given a domain containing $2^k$ values, including zero, we can construct a new half-size domain $D'$ by taking
+$x \cdot (x+k)$. because the function $f(x) = x \cdot (x+k)$ returns the same value for $x$ and $x + k$.
+
+| $x$               | 0   | 1   | 2   | 3   | 4   | 5   | 6   | 7   | 8   | 9   | 10  | 11  | 12  | 13  | 14  | 15  |
+| ----------------- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
+| $x \cdot (x + 1)$ | 0   | 0   | 6   | 6   | 7   | 7   | 1   | 1   | 4   | 4   | 2   | 2   | 3   | 3   | 5   | 5   |
+
+Let us explain this table:
+
+- With $x = 0$: $x + 1 = 1$ $\to$ $x \cdot (x + 1) = 0 \cdot 1 = 0$
+- With $x = 1$: $x + 1 = 1 + 1 = 0 \to x \cdot (x+1) = 1 \cdot 0 = 0$
+- ...
+- With $x = 15 \iff x = 1111$; $x + 1 = 1111 + 0001 = 1110 = 14$ $\to x \cdot (x + 1) = 1111 \cdot 1110 = 0101 = 5$
+
+So, to do FFT on binary field, we should convert $f(x)$ into:
+$$f(x) = evens(x \cdot (k-x)) + x \cdot odds(x \cdot (k-x))$$
+And $f(x + k) = evens((x + k) \cdot (-x)) + (x + k) \cdot odds((x+k) \cdot (-x))$
+$= evens(x \cdot (k-x)) + (x + k) \cdot odds(x \cdot (k-x))$.
+
+To convert $f(x)$ to $evens$ and $odds$ above in $O(N \cdot logN)$, we can use the
+[freshman's dream](https://en.wikipedia.org/wiki/Freshman%27s_dream) theory that in binary field:
+$$(x + y)^2 = x^2 + y^2$$
+Then, do the same things as in the FFT section.
+
+Here is the code for all the implementation that we mentioned above:
+[link](https://github.com/ethereum/research/tree/master/binary_fft).
+
+## Reference
+
+[Fast Fourier Transforms](https://vitalik.eth.limo/general/2019/05/12/fft.html)
+<br>
+Vitalik Buterin
```

Comment:
If Vitalik's article was already linked above, there's no need to mention his name again